postgresql/sql-scripts/citydb-pkg/delete.sql

```diff
@@ -146,6 +150,129 @@ END;
 $body$
 LANGUAGE plpgsql STRICT;
 
+/******************************************************************
+* terminate feature based on an id array
+******************************************************************/
+CREATE OR REPLACE FUNCTION citydb_pkg.terminate_feature(pid_array bigint[], metadata JSON DEFAULT '{}') RETURNS SETOF BIGINT AS
+$body$
+DECLARE
+  terminated_ids bigint[] := '{}';
+  child_feature_ids bigint[] := '{}';
+  termination_timestamp TIMESTAMP;
+  update_reason TEXT;
+  update_person TEXT DEFAULT USER;
+  udpate_lineage TEXT;
+BEGIN
+  termination_timestamp = now();
+
+  IF metadata IS NOT NULL THEN
+    IF metadata->>'reason_for_update' IS NOT NULL THEN
+      update_reason = metadata->>'reason_for_update';
+    END IF;
+
+    IF metadata->>'updating_person' IS NOT NULL THEN
+      update_person = metadata->>'updating_person';
+    END IF;
+
+    IF metadata->>'lineage' IS NOT NULL THEN
+      udpate_lineage = metadata->>'lineage';
+    END IF;
+  END IF;
+
+  WITH terminated_objects AS (
+    UPDATE
+       feature f
+    SET
+       termination_date = termination_timestamp,
+       last_modification_date = termination_timestamp,
+       reason_for_update = update_reason,
+       updating_person = update_person,
+       lineage = udpate_lineage
```

Comment:
I suggest using the following update statements (lines 186-190):

```sql
termination_date = COALESCE((metadata->>'termination_date')::timestamp with time zone, now()),
last_modification_date = COALESCE((metadata->>'last_modification_date')::timestamp with time zone, now()),
reason_for_update = COALESCE(metadata->>'reason_for_update', f.reason_for_update),
updating_person = COALESCE(metadata->>'updating_person', USER),
lineage = COALESCE(metadata->>'lineage', f.lineage)
```

The advantage is that exisiting values for `reason_for_update`, `updating_person`, and `lineage` are not overwritten with NULL. Moreover, the user can also provide values for `termination_date` and `last_modification_date`.

Reply:
Addition to my above comment: Instead of using now() for both `termination_date` and `modification_date`, we should maybe use a local variable `timestamp = now()` and then use this one in COALESCE:

```
termination_date = COALESCE((metadata->>'termination_date')::timestamp with time zone, timestamp),
last_modification_date = COALESCE((metadata->>'last_modification_date')::timestamp with time zone, timestamp),
```

This way,  `termination_date` and `modification_date` receive the same time by default.

Reply:
Using now() multiple times is safe: https://www.postgresql.org/docs/15/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT